### PR TITLE
KAFKA-8256: Replace Heartbeat request/response with automated protocol

### PR DESCRIPTION
**Pull request by mimaison**
### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/common/requests/HeartbeatRequest.java [hunk 1]
```diff
@@ -16,108 +16,76 @@
...
-                return new HeartbeatResponse(throttleTimeMs, Errors.forException(e));
+                response.setThrottleTimeMs(throttleTimeMs);
+                return new HeartbeatResponse(response);
             default:
                 throw new IllegalArgumentException(String.format("Version %d is not valid. Valid versions for %s are 0 to %d",
                         versionId, this.getClass().getSimpleName(), ApiKeys.HEARTBEAT.latestVersion()));
         }
     }
 
     public String groupId() {
```
Comment (abbccdda) [inline clients/src/main/java/org/apache/kafka/common/requests/HeartbeatRequest.java]: We could remove this function

File: clients/src/main/java/org/apache/kafka/common/requests/HeartbeatRequest.java [hunk 2]
```diff
@@ -16,108 +16,76 @@
...
-        return groupId;
+        return data.groupId();
     }
 
     public int groupGenerationId() {
```
Comment (abbccdda) [inline clients/src/main/java/org/apache/kafka/common/requests/HeartbeatRequest.java]: Same for generation and member.id. We could just call `request.data().generationId()`

File: clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java [hunk 3]
```diff
@@ -1439,7 +1440,9 @@ public void shouldAttemptToRejoinGroupAfterSyncGroupFailed() throws Exception {
             public boolean matches(AbstractRequest body) {
                 return true;
             }
-        }, new HeartbeatResponse(Errors.REBALANCE_IN_PROGRESS), coordinator);
+        }, new HeartbeatResponse(
+                new HeartbeatResponseData().setErrorCode(Errors.REBALANCE_IN_PROGRESS.code())),
+            coordinator);
```
Comment (abbccdda) [inline clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java]: nit: format

**Review by abbccdda — COMMENTED** [review]
Thanks for the PR! Some minor comments.

**Comment by abbccdda** [comment]
@mimaison Do you still work on this PR?

**Comment by mimaison** [comment]
Yes but I'm at the Kafka Summit at the moment. Thanks for the review, I'll
take a look when I get back

**Comment by abbccdda** [comment]
@mimaison Perfect, enjoy the talks!

File: clients/src/main/java/org/apache/kafka/common/requests/HeartbeatRequest.java [hunk 4]
```diff
@@ -16,108 +16,76 @@
...
-                append(")");
-            return bld.toString();
+            return data.toString();
         }
     }
 
-    private final String groupId;
-    private final int groupGenerationId;
-    private final String memberId;
+    private final HeartbeatRequestData data;
```
Comment (abbccdda) [inline clients/src/main/java/org/apache/kafka/common/requests/HeartbeatRequest.java]: We could make this field access `public`

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 5]
```diff
@@ -1414,7 +1414,8 @@ class KafkaApis(val requestChannel: RequestChannel,
     // the callback for sending a heartbeat response
     def sendResponseCallback(error: Errors) {
       def createResponse(requestThrottleMs: Int): AbstractResponse = {
-        val response = new HeartbeatResponse(requestThrottleMs, error)
+        val response = new HeartbeatResponse(
+            new HeartbeatResponseData().setThrottleTimeMs(requestThrottleMs).setErrorCode(error.code))
```
Comment (abbccdda) [inline core/src/main/scala/kafka/server/KafkaApis.scala]: nit: format one parameter per line

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 6]
```diff
@@ -1424,7 +1425,8 @@ class KafkaApis(val requestChannel: RequestChannel,
 
     if (!authorize(request.session, Read, Resource(Group, heartbeatRequest.groupId, LITERAL))) {
       sendResponseMaybeThrottle(request, requestThrottleMs =>
-        new HeartbeatResponse(requestThrottleMs, Errors.GROUP_AUTHORIZATION_FAILED))
+        new HeartbeatResponse(
+            new HeartbeatResponseData().setThrottleTimeMs(requestThrottleMs).setErrorCode(Errors.GROUP_AUTHORIZATION_FAILED.code)))
```
Comment (abbccdda) [inline core/src/main/scala/kafka/server/KafkaApis.scala]: Same here

File: core/src/test/scala/unit/kafka/server/RequestQuotaTest.scala [hunk 7]
```diff
@@ -298,7 +298,7 @@ class RequestQuotaTest extends BaseRequestTest {
           )
 
         case ApiKeys.HEARTBEAT =>
-          new HeartbeatRequest.Builder("test-group", 1, "")
+          new HeartbeatRequest.Builder(new HeartbeatRequestData().setGroupId("test-group").setGenerationid(1).setMemberId(""))
```
Comment (abbccdda) [inline core/src/test/scala/unit/kafka/server/RequestQuotaTest.scala]: nit: use `JoinGroupRequest.UNKNOWN_MEMBER_ID` instead of `""`

**Review by hachikuji — APPROVED** [review]
LGTM.  I will merge after the comments from @abbccdda are addressed and the patch is rebased.

**Review by abbccdda — APPROVED** [review]
Thank you for the updates! LGTM

**Pull request merged**